viewing_party/party.py

```diff
@@ -1,23 +1,257 @@
 
+def get_unique_watched(user_data):
+
+    copy_from_user_data = copy.deepcopy(user_data)
+
+    friends_movie_titles = []
+    users_movie_titles = []
+    my_list = copy_from_user_data["watched"]
+    friends_list = copy_from_user_data["friends"]
+    
+    for i in range(len(friends_list)):
+        values_of_watched = friends_list[i]["watched"]
         
+        for j in range(len(values_of_watched)):
+            titles = values_of_watched[j]["title"]
+            friends_movie_titles.append(titles)
+
+    for i in range(len(my_list)):
+        users_movie_titles.append(my_list[i]["title"])
+
+    user_unique_movies = set(users_movie_titles) - set(friends_movie_titles)
+    users_unique_movies_list = list(user_unique_movies)
+    
+    result_list = []
+
+    for movie_title in users_unique_movies_list:
+        for movie in my_list:
+            if movie["title"] == movie_title:
+                result_list.append(movie)
+
+    return result_list
+
+
+def get_friends_unique_watched(user_data):
+
+    copy_from_user_data = copy.deepcopy(user_data)
+    
+    friends_movie_titles = []
+    users_movie_titles = []
+    friends_movies = []
+    my_list = copy_from_user_data["watched"]
+    friends_list = copy_from_user_data["friends"]
+    
+    for i in range(len(friends_list)):
+        values_of_watched = friends_list[i]["watched"]
+        
+        for j in range(len(values_of_watched)):
+            titles = values_of_watched[j]["title"]
+            friends_movie_titles.append(titles)
+            friends_movies.append(values_of_watched[j])
+
+    for i in range(len(my_list)):
+        users_movie_titles.append(my_list[i]["title"])
+
+    friends_unique_movies = set(friends_movie_titles) - set(users_movie_titles)
+    friends_unique_movies_list = list(friends_unique_movies)
+    
+    result_list = []
+
+    for movie_title in friends_unique_movies_list:
+        for movie in friends_movies:
+            if movie["title"] == movie_title:
+                result_list.append(movie) 
+                break
+
+    return result_list
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    friends_list = user_data["friends"]
+    user_watched = user_data["watched"]
+    user_watched_titles = []
+
+    # Iterate through user's watched dicts and get the titles user has watched
+    for i in range(len(user_watched)):
+        user_watched_titles.append(user_watched[i]["title"])
+
+    movie_recs = []
+    for dict in friends_list:
+        for value in dict.values():
+            for movie in value:
+                friends_title = movie["title"]
+
+                for user_dict in user_watched:
+                    if friends_title not in user_watched_titles and movie["host"] in user_data["subscriptions"] and movie not in movie_recs:
+                        movie_recs.append(movie)
+    
+    return movie_recs

```

Comment:
A quadruple loop! I highly recommend trying to see if we can reduce the number of nested loops. Tools like VS Code Debugger can be really helpful in visualizing the shape of the data we're working with.